### DISCUSSION TITLE
How to fetch data in Client Components, the NextJS way

### DISCUSSION THREAD

**Discussion by MankiratSingh1315** [vercel/next.js#68158] [Help]
### Summary

I was working on one of my project with NextJS 14.2 and I want to fetch and display dynamic data to the user

Below I have listed some of my approaches but none of them worked - 
1. Firstly I found server actions and used them with async components to fetch the data
    - I implemented it as follows
      ```ts
      import { getListings } from @/actions/listings

      export default async function Listngs() {
          const listings: Listing[] = await getListings();
          return (
              <div>
                  {/* Listing cards rendering */}
              </div>
          );
      }
      ```
       But later I found that I misunderstood the documentation as it says server actions are only for submitting form-data(I think so) and on building the application it fetches the data from database while building only and later on views it as a static page only so I moved to some other method

2. Then I found [this](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-server-with-fetch) piece of documentation that says make a next js GET route handler to process your data, fetch the API from the server only and it will view the page on the frontend server side rendered only.
    - I implemented it as follows
      ```ts
      async function getListings(){
        const response = await fetch(`${process.env.NEXT_PUBLIC_CLIENT_URL}/api/listings/`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
                cache: 'no-store'
            });
            const yoyo = await response.json();
            console.log(yoyo);
            return yoyo;
        }

      export default async function Listngs() {
          const listings: Listing[] = await getListings();
          return (
              <div>
                  {/* Listing cards rendering */}
              </div>
          );
      }
      ```
       But still I am having the same problem that it works in development environment and when I build it, it still don't renders data in production environment.

Can someone help me out with this?
My main question is that the only way to do this is useEffect method in client components to get the data?
If yes then why the documentation don't says so?

I found [this](https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#parallel-data-fetching) documentation for parallel fetching of data also but it also don't work anyhow.

**Comment by icyJoseph** [comment]
- https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-client-with-route-handlers
- https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-client-with-third-party-libraries

Yeah, in client components, you _kind of_ have to still use the, render then fetch pattern (a use-effect that executes and saves the result to a stateful variable), but you can use 3rd party libraries in combination with Route Handlers, that can make that a better experience.

Alternatively, and the reason why I say, _kind of_, is that, you can, in Next.js (or in React canary rather), use the `use` hook https://react.dev/reference/react/use#noun-labs-1201738-(2) to consume a promise, but promise references ought to be stable. 

And yet, another cool way is that, a server component, can pass a promise to a client component as props, and the client component can `use(thatPromiseProp)`, and suspend while the response streams to the browser! https://react.dev/reference/react/use#streaming-data-from-server-to-client

**Reply by MankiratSingh1315** [reply]
The page I am trying to render data on is a server component and it needs no client interaction
Just the data to be viewed to the client is dynamic

**Reply by icyJoseph** [reply]
Post a pseudo-code example. I think we are about to talk past each other otherwise.

**Reply by MankiratSingh1315** [reply]
Like the one given below, the `Listings` component here is a server component because it is an async component which fetches the data from the route handler
```ts
async function getListings(){
  const response = await fetch(`${process.env.NEXT_PUBLIC_CLIENT_URL}/api/listings/`, {
          method: 'GET',
          headers: {
              'Content-Type': 'application/json',
          },
          cache: 'no-store'
      });
      const yoyo = await response.json();
      console.log(yoyo);
      return yoyo;
  }

export default async function Listings() {
    const listings: Listing[] = await getListings();
    return (
        <div>
            {/* Listing cards rendering */}
        </div>
    );
}
```

I am unable to get data on the page upon building my application

I used a similar approach on some other component, just the thing was that the parent component was an async component as given above and it passes the data to a child component which is client component, and it works fine as given below
```ts
// Parent.ts
const getApprovedListings = async () => {
    const response = await fetch(`${process.env.NEXT_PUBLIC_CLIENT_URL}/api/listings/getApprovedLisitngs`, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
        cache: 'no-store'
    });
    const data = await response.json();
    return data;
}

export default async function ParentComponent() {
    const listings = await getApprovedListings();
    return <ChildComponent data={listings} />;
}
```

```ts
// Child.ts
"use client"
export default const ChildComponent = ({ data }: any) => {
    return (
        <div>{/* rendering data */}</div>
    );
}
```

but the Listings component I am trying to make is all a server component

**Reply by icyJoseph** [reply]
- Alright, let us see.

First, fetching from Route Handlers from server components is not good, and for pages rendered at Build time, won't even work.

- https://nextjs-faq.com/fetch-api-in-rsc

The link above has more info, but you should just get the data straight up from its source. 

During build time there's no server to listen for your request. And on demand runtime rendering would do an extra unnecessary round-trip out to the internet and back to your server.

With that being said, I'm still not understanding your issue, but maybe the text above already helps you out?


If everything is a server component on that - page, including th page itself, then it should work at least on dev mode. Are you sure there's not a client component parent in the tree, owning this component by accident?

**Reply by MankiratSingh1315** [reply]
The text is helpful but does it mean the first approach I mentioned in the original question was correct?
Because the function which was imported from `@/actions/listings` have all the required logic...
I mean I am using server function in server component so there should be no problem?

and yes it works fine in dev mode problem occurs when I build the component

**Comment by Juneezee** [comment]
> My main question is that the only way to do this is useEffect method in client components to get the data?

There are many data fetching libraries that you can use. Some examples are:

1. Tanstack Query (https://tanstack.com/query/latest/docs/framework/react/quick-start)
2. RTK Query (https://redux-toolkit.js.org/rtk-query/overview)
3. swr (https://swr.vercel.app/)

> I found that I misunderstood the documentation as it says server actions are only for submitting form-data

You can definitely fetch data using Server Actions. Personally, I think it is cleaner to do it this way. Here is what I am doing in my project using Server Actions + Tanstack Query

```ts
// users/_actions/listUsers.ts
"use server";

export default async function listUsers(): Promise<User> {
  // ...
}


// users/page.tsx
"use client";

export default function UsersPage(): React.JSX.Element {
  const {
    data: listUsersQueryData,
    error: listUsersQueryError,
    isPending: isListUsersQueryPending,
    refetch: refetchListUsersQuery,
  } = useQuery<User[]>({
    queryKey: ["User", "LIST"],
    queryFn: () => listUsers().then(rejectApiErrorResponse),
  });

  // ...
}
```

I store all my Server Actions code in a [private folder](https://nextjs.org/docs/app/building-your-application/routing/colocation#private-folders) `_actions`.

**Reply by icyJoseph** [reply]
One caveat straight up from the docs though:

- https://react.dev/reference/rsc/use-server#caveats

> Server Actions are designed for mutations that update server-side state; they are not recommended for data fetching. Accordingly, frameworks implementing Server Actions typically process one action at a time and do not have a way to cache the return value.

I guess, if it works it works, but good to be aware that the docs don't recommend it. And yeah, sometimes, breaking some rules makes sense, as long as the consequences are clearly understood.

**Reply by MankiratSingh1315** [reply]
what if the UserPage component is a server component and I don't want to put "use client" there?

**Reply by Juneezee** [reply]
> what if the UserPage component is a server component and I don't want to put "use client" there?

If your `page.tsx` is a server component, then you can fetch the data directly without the hooks.

**Reply by MankiratSingh1315** [reply]
> If your `page.tsx` is a server component, then you can fetch the data directly without the hooks.

So the first approach in my question which I thought to be misunderstood was correct and should have no problem right?

**Reply by Juneezee** [reply]
> > If your `page.tsx` is a server component, then you can fetch the data directly without the hooks.
> 
> So the first approach in my question which I thought to be misunderstood was correct and should have no problem right?

@MankiratSingh1315 Basically yes. 

There is another tip. If your `Listing[]` comes from your database, you can fetch from your DB client directly and don't need an extra Route Handler. For example:

```tsx
export default async function Listngs() {
    const listings: Listing[] = await db.findAllListings();
    return (
        <div>
            {/* Listing cards rendering */}
        </div>
    );
}
```

**Reply by MankiratSingh1315** [reply]
The problem I am facing is that either I get listings da directly from database or use a function, everything works fine in dev mode(I can add new listings in the database and it is visible on frontend after refresh) but when I build my application, new listings and any of the changes in listings database does not appear on the frontend

**Reply by MankiratSingh1315** [reply]
[NextJS Learn Fetching data guide](https://nextjs.org/learn/dashboard-app/fetching-data#fetching-data-for-revenuechart)
I want to do dynamic rendering of data which is not given in this guide, could you please provide a sample code for it?
